src/lib.rs

```diff
@@ -1,9 +1,361 @@
+    ///
+    /// # Sharing
+    ///
+    /// Once the connection is established and we know the other ends peer ID
+    /// the repo will query all the documents it knows about using `Store::list` and
+    /// then send sync messages for every document for which
+    /// `Share::should_share` returns `ShareDecision::Share`
+    pub async fn connect<RecvErr, SendErr, Source>(
+        self,
+        source: Source,
+        stream: Store,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::error::Error,
+        SendErr: std::error::Error,
+        Store: Sink<Message, Error = SendErr> + Stream<Item = Result<Message, RecvErr>>,
+        Source: std::fmt::Debug + Eq,
+    {
+        todo!()
+    }
+
+    /// Connect a tokio io object
+    ///
+    /// This implements a simple length prefixed framing protocol and is intended for use with
+    /// stream oriented transports
+    #[cfg(feature = "tokio")]
+    pub async fn connect_tokio_io<Io, Source>(
+        self,
+        source: Source,
+        io: Io,
+        direction: ConnDirection,
+    ) -> Result<(), error::RunConnection<error::CodecError, error::CodecError>>
+    where
+        Io: tokio::io::AsyncRead + tokio::io::AsyncWrite,
+    {
+        todo!()
+    }
+
+    /// Attempt to load the document given by `id` from storage
+    ///
+    /// If the document is not found in storage then `Ok(None)` is returned.
+    ///
+    /// Note that this _does not_ attempt to fetch the document from the
+    /// network.
+    pub async fn load(&self, id: DocumentId) -> Result<Option<DocHandle>, Store::Error> {
+        todo!()
+    }
+
+    /// Request a document from a peer in the network
+    ///
+    /// This will connect to the peers specified by `peers` and run the sync
+    /// protocol for the given document ID. Once the sync protocol has run (or
+    /// not as the case may be) this function will return the document handle.
+    ///
+    /// This can be used both to fetch a document you don't have and to request
+    /// updates you think another peer has but which you haven't been sent yet.
+    ///
+    /// # Arguments
+    ///
+    /// * id - The ID of the document to request
+    /// * peer - The peer(s) to request the document from.
+    pub async fn request(
+        &self,
+        id: DocumentId,
+        peers: PeerSpec,
+    ) -> Result<Option<DocHandle>, error::Request> {
+        todo!()
+    }
+
+    /// Create a new document, store it, and return a `DocHandle` for it
+    ///
+    /// In addition to creating the document this will also enqueue sync messages
+    /// to every connected peer for whom `Share::should_share` returns `ShareDecision::Share`
+    ///
+    /// # Errors
+    ///
+    /// * If a document with this ID already exists in storage
+    pub async fn create(&self, doc: am::Automerge) -> Result<DocHandle, error::Create<Store>> {
+        todo!()
+    }
```

Comment:
I like the share policy idea, and I think it can be added at a later stage, where a first version simply syncs to all connected peers. Note that a "share policy" could be achieved by running multiple repos, I think.